### PR TITLE
Add setting for leniency on Gson instance.

### PR DESCRIPTION
**Pull request by JakeWharton**
Add a JsonReader factory (for parity with the JsonWriter one) which provides a configured instance using the Gson settings.

### CODE REVIEW

**Comment by swankjesse** [comment]
This is good. It’s a bit unfortunate that by default `Gson` is actually lenient in some cases and not-lenient in others. Oh well.

**Comment by swankjesse** [comment]
LGTM. @inder123 any thoughts?

**Comment by JakeWharton** [comment]
> It’s a bit unfortunate that by default Gson is actually lenient in some cases and not-lenient in others. Oh well.

Agree! Too late to change. Just trying to make the best of that unfortunate situation.

File: gson/src/test/java/com/google/gson/functional/LeniencyTest.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  private Gson gson;
+
+  @Override
+  protected void setUp() throws Exception {
+    super.setUp();
+    gson = new GsonBuilder().setLenient().create();
+  }
+
+  public void testLenientFromJson() {
+    String json = gson.fromJson("\"Hello\"", String.class);
```
Comment (swankjesse) [inline gson/src/test/java/com/google/gson/functional/LeniencyTest.java]: Wanna use another example? It’s likely we’ll support top-level primitives.

Comment (JakeWharton) [inline gson/src/test/java/com/google/gson/functional/LeniencyTest.java]: Yep yep. I was planning on porting the RFC 7159 change to here today anyway.

Comment (JakeWharton) [inline gson/src/test/java/com/google/gson/functional/LeniencyTest.java]: Updated to use comments and single-quote string in otherwise valid JSON.

**Comment by swankjesse** [comment]
LGTM!

File: gson/src/main/java/com/google/gson/Gson.java [hunk 2]
```diff
@@ -705,6 +714,15 @@ public JsonWriter newJsonWriter(Writer writer) throws IOException {
   }
 
   /**
+   * Returns a new JSON writer configured for the settings on this Gson instance.
+   */
+  public JsonReader newJsonReader(Reader reader) {
```
Comment (inder123) [inline gson/src/main/java/com/google/gson/Gson.java]: The method body seems fairly straightforward. Why can't the developer do this themselves?
We could just expose an isLenient() method instead.

Comment (JakeWharton) [inline gson/src/main/java/com/google/gson/Gson.java]: Parity with `newJsonWriter` and protection against future configuration options being added.

Comment (inder123) [inline gson/src/main/java/com/google/gson/Gson.java]: Ok, that seems reasonable.

**Pull request merged**